### ISSUE TITLE
Add alias support to version automation

### ISSUE THREAD

**Issue by agjohnson** [rtfd/readthedocs.org#5318]
In #4001 we started by adding version automation rules to enable new versions.

In this next phase, we will address the use case of automation of version aliases. Prior to this, we might need to refactor how we are managing version alias, and potentially more widely support alias management in our UI.

Example use cases here are:

* User want to set up aliases for minor versions that point to bugfix releases. For instance, a project has `1.0`, `1.1`, and `2.0` versions active in docs. On release of `2.0.5`, the `2.0` alias will automatically repoint to this new tag
* User has an odd branch naming convention, wants to automate adding versions derived from these branch names. For instance, project repository uses major version branches for backporting use and wants `version-3.0` to show up as `3.0` automatically.

Some rule example psuedo logic are:

* On [Regex match /2\.1\.[\d+]/], [Set alias `2.1`]
* On [Regex match /([\d+])\.([\d+])\.[\d+]/], [Set alias `{match[0]}.{match[1]}`]

### Work

- [ ] Add an action "Set alias". This action supports some templating, such as the docker hub version automation -- `{major}`, `{minor}`, `{\1}` (regex grouping), etc
- [ ] Add UI pieces for rules, if anything new is required in our UI

**Comment by stsewd** [comment]
Another use case is for users that have versions not compatible with pep 440, instead of renaming all of their versions, they can just rename the slug on rtd and automate this with automation rules. We want to allow users to edit the version slug first.

**Comment by stsewd** [comment]
> This seems like the right fix for my https://github.com/dagger/dagger/issues/3767, let me know if this is true.
> Would this feature allow extracting the semver from the tag and use that?

Yes, if implemented, this would solve your problem. Thanks for commenting your use case!

**Comment by stsewd** [comment]
@thibaudcolas we will likely first allow users to change the slug or create an alias manually https://github.com/readthedocs/readthedocs.org/issues/9684 before adding an automation rule. But these issues are still pending some decisions, you can request changing your version slug by emailing us to support, just note that any existing version with the old slug will 404, and you need to trigger a new build on the branch/tag that was renamed.

**Comment by stsewd** [comment]
Hi, we now allow changing the slug of versions. Some notes:

- Changing the slug of an active version will result on its previous documentation being deleted, and a new build being triggered. So be careful when renaming active versions.
- You'll probably want to create a redirect from the old version to the new one (an exact redirect with a wildcard should do it).
- You may still see the old name in some places, as changing the slug only affects the URL and how the APIs identify that version. We are planning on introducing a new field that will control how the version is shown to users and also use that value when sorting the versions (https://github.com/readthedocs/readthedocs.org/issues/11979).
- Sorting of versions makes use of the slug, but I consider that to be more of a bug (https://github.com/readthedocs/readthedocs.org/issues/11182), so that may change in the future. But with https://github.com/readthedocs/readthedocs.org/issues/11979 you should have control over that in the future.
- You can't change the slug of the latest and stable versions that are managed by RTD.

This feature is under a feature flag, if you want to try it out let me know the projects you want to test this on.

**Comment by Jackenmen** [comment]
> You can't change the slug of the latest and stable versions that are managed by RTD.

Does this mean that this feature won't _yet_ allow, say, to have a branch named `stable-docs` and have it appear under `project.readthedocs.io/en/stable/`?

**Comment by stsewd** [comment]
> Does this mean that this feature won't yet allow, say, to have a branch named stable-docs and have it appear under project.readthedocs.io/en/stable/?

Good question. Haven't tested it yet, but it should be possible, but not straightforward to do. If RTD already created a version named `latest` or `stable` automatically (marked as machine created), then you'll need to do a couple of steps to remove the "machine" mark from those, and then rename latest/stable to something else, and then rename your other version to use that name.

Since we already allow users to override stable/latest by creating a branch/tag with that name, I think it's valid to allow doing the change from the form, it would be required for user to first disable the RTD latest/stable version, or maybe do that automatically, but show a big warning.

**Comment by stsewd** [comment]
@thibaudcolas done!

**Comment by thibaudcolas** [comment]
Finally got around to trying this – it seems to already work better than I was expecting!

See for example Wagtail’s [`6.4` slug](https://docs.wagtail.org/en/6.4/), which used to be [stable-6.4.x](https://docs.wagtail.org/en/stable-6.4.x/), which is a branch build of [stable/6.4.x](https://github.com/wagtail/wagtail/tree/stable/6.4.x) in GitHub. Proceeding with this would allow us to transition away from tag builds, replacing [v6.4](https://docs.wagtail.org/en/v6.4/) and other tags up to v6.4.3 with a single docs build with a stable slug.

And it’s not just in the URL, it also looks like what we were expecting in the Addons:

![Addons picker for Wagtail](https://github.com/user-attachments/assets/5a983fd1-d92e-4d6b-8eb8-e6285f0612db)

---

I wasn’t able to test the sort order of the versions in this list yet. In our case we would want "latest stable \[version-specific branch and tags builds, mixed together, in the order of the version number\]". I’ll need to try this out on sphinx-wagtail-theme.

**Comment by agjohnson** [comment]
I keep handling support requests where automation and version custom slugs would be lovely, I'd like to try to prioritize this once we're done with some larger projects.

The big questions to me are around UX still. One is that I don't quite know how we manage moving custom slugs around in the case like:

- Branch `1.2.3` is released
- Automation rule gives version `1.2.3` custom slug `1.2`
- Branch `1.2.4` is released
- (?) The custom slug needs to be removed from all previous versions first
- Automation rule gives version `1.2.4` custom slug `1.2`

Just an example, there are other use cases besides semver tagging like this where the custom slug would move between versions automatically.

This is maybe the "alias" version idea, though custom slugs and aliases together feel quite confusing.

**Comment by humitos** [comment]
> (?) The custom slug needs to be removed from all previous versions first

This makes sense. However, it will be _one_ version with `1.2` custom slug since we can't have two versions with the same slug.

**Comment by stsewd** [comment]
I guess there are two options in this use case:

- Would 1.2 exist together with 1.2.3? If so, that would be an alias, not a rename. So users are better off doing that aliasing at the git level (just another tag/branch).
- If any 1.2.x version is renamed to be 1.2, then there will be a downtime during the rename. As 1.2 won't be available until the latest 1.2.x version finishes building. Which again, would be easier for users to do this at the git level. I think renaming slugs is more to cover use cases where users need to do a single rename, like omitting or adding a prefix/suffix to the version.

**Comment by agjohnson** [comment]
Yeah if the user needs all these versions active they'd probably be better off right now to use dedicated major/minor release series branches. In this case the users just wanted to activate the `1.2.3` version as the new `1.2` version, not activate both.

Maintaining "version aliases" strictly as pointers to versions could probably exist alongside editable versions. I know we've talked about both, though it certainly makes our serving logic more complicated. But overall this feature compliments well and version aliases are a cleaner way to describe our built in `latest`/`stable` too. 

I noted above but this semver example was just one example of the need for this feature. Even just the basic feature of altering tags like `releases/v.3.1.0` to a version `3.1.0` feels like something we should support through automation rules.

> If any 1.2.x version is renamed to be 1.2, then there will be a downtime during the rename. As 1.2 won't be available until the latest 1.2.x version finishes building

This sounds like a problem with automation rules in general, not with this feature.

We apply automation rules before the build, in the webhook -- correct?. So there is a period where any change to a version that has not been built yet will result in a period of downtime -- ie changing the default version to an unbuilt version.

**Comment by agjohnson** [comment]
Just an update here, we discussed this issue again.

We're still not sure if we want to take on the complexity required for this feature yet. We'll wait on adding this feature for now, but we'll at least make this use pattern possible by making sure our API can allow version editing. This should allow users to at least script a solution.

We might take this feature on if we also end up needing similar technical changes for:

- #5319 
- #12128

**Comment by humitos** [comment]
> we'll at least make this use pattern possible by making sure our API can allow version editing. This should allow users to at least script a solution.

We have an issue for this, and it's already in the roadmap for Q4

- https://github.com/readthedocs/readthedocs.org/issues/11970
